pkg/mic/mic.go

```diff
@@ -401,57 +441,166 @@ func (c *Client) makeAssignedIDs(azID *aadpodid.AzureIdentity, azBinding *aadpod
+			glog.Errorf("Lookup of node %s resulted in error %v", n, err)
+			continue
+		}
+		idList, err := c.getUserMSIListForNode(node)
+		if err != nil {
+			glog.Errorf("Getting list of msi's from node %s resulted in error %v", n, err)
+		}
+		nodeIdentityList[n] = idList
+	}
+
+	isNodeErrored := func(node string) bool {
+		for _, n := range nodesWithError {
+			if n == node {
+				return true
 			}
-			return err
 		}
+		return false
 	}
-	return nil
-}
 
-func (c *Client) removeAssignedIDsWithDeps(assignedID *aadpodid.AzureAssignedIdentity, inUse bool) error {
-	err := c.CRDClient.RemoveAssignedIdentity(assignedID)
-	if err != nil {
-		glog.Error(err)
-		return nil
+	if addList != nil && len(*addList) > 0 {
+		for _, createID := range *addList {
+			id := createID.Spec.AzureIdentityRef
+			binding := createID.Spec.AzureBindingRef
+			isUserAssignedMSI := c.checkIfUserAssignedMSI(id)
+
+			idExistsOnNode := c.checkIfMSIExistsOnNode(id, createID.Spec.NodeName, nodeIdentityList)
+			nodeHasErrors := isNodeErrored(createID.Spec.NodeName)
+
+			if !nodeHasErrors || idExistsOnNode {
+				// the identity was successfully assigned to the node
+				c.EventRecorder.Event(binding, corev1.EventTypeNormal, "binding applied",
+					fmt.Sprintf("Binding %s applied on node %s for pod %s", binding.Name, createID.Spec.NodeName, createID.Name))
+
+				glog.V(5).Infof("Initiating assigned id creation for pod - %s, binding - %s", createID.Spec.Pod, binding.Name)
+
+				if isUserAssignedMSI {
+					err := c.createAssignedIdentity(&createID)
+					if err != nil {
+						// Since k8s event has only Info and Warning, using Warning.
+						c.EventRecorder.Event(binding, corev1.EventTypeWarning, "binding apply error",
+							fmt.Sprintf("Creating assigned identity binding %s node %s for pod %s resulted in error %v", binding.Name, createID.Spec.NodeName, createID.Name, err))
+						glog.Error(err)
+					}
+				}
+			}
+		}
 	}
-	if !inUse {
-		id := assignedID.Spec.AzureIdentityRef
-		if id.Spec.Type == aadpodid.UserAssignedMSI {
 
-			node, err := c.NodeClient.Get(assignedID.Spec.NodeName)
-			if err != nil {
-				return err
+	if deleteList != nil && len(*deleteList) > 0 {
+		for _, delID := range *deleteList {
+			inUse := c.checkIfInUse(delID, newAssignedIDs)
+			id := delID.Spec.AzureIdentityRef
+			removedBinding := delID.Spec.AzureBindingRef
+			isUserAssignedMSI := c.checkIfUserAssignedMSI(id)
+			nodeHasErrors := isNodeErrored(delID.Spec.NodeName)
+
+			idExistsOnNode := c.checkIfMSIExistsOnNode(id, delID.Spec.NodeName, nodeIdentityList)
+
+			if !nodeHasErrors || !idExistsOnNode {
+				// the identity was successfully removed from node
+				c.EventRecorder.Event(removedBinding, corev1.EventTypeNormal, "binding removed",
+					fmt.Sprintf("Binding %s removed from node %s for pod %s", removedBinding.Name, delID.Spec.NodeName, delID.Spec.Pod))
+
+				// remove assigned identity crd from cluster
+				if err := c.removeAssignedIdentity(&delID); err != nil {
```

Comment:
Similarly, removing the assigned identity only when the identity has been successfully removed from the node